feat: 1inch swapper pre-release second pass fixes

woodenfurniture opened the pull request:
## Description

Second pass fixes for 1inch - note this does not yet enable it on prod since ops didn't have a chance to test this yet with the current gas spike.

~~Enables 1inch swapper for all environments.~~

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [x] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

NA

## Risk

Risk of broken and/or incorrect trades when 1inch swapper is being used. This includes but isnt limited to:
- incorrect amount estimates
- incorrect gas estimates
- incorrect amounts being received
- incorrect assets being received
- app crashes during trades
- dead clicks during trades
- failed transactions
- etc

## Testing

Have tested quotes including gas estimates and receive amounts. Only tested 1 trade due (FOX -> MATIC) to extremely high gas prices and lack of balance across various eth chain assets.

### Engineering

See above

### Operations

See above. Also handy to note for testing, this swapper only uses erc20 assets on ethereum mainnet. That means ETH is excluded and will not produce a valid trade pair when selected.

## Screenshots (if applicable)

0xApotheosis reviewed (commented):
First pass when testing:

Approval does not seem to be working with 1inch.

I successfully executed an unlimited approval TX with 1inch on the FOX contract (https://etherscan.io/tx/0x7e8497ea260722729bc0697c593e71eee0b17ad31b4db85796f912b080113d34), but the TX was never picked up and shown as complete in the UI:

<img width="474" alt="Screenshot 2023-05-05 at 11 49 46 am" src="https://user-images.githubusercontent.com/97164662/236363083-1d15b092-a8ed-4160-b786-10fa06366752.png">

And when trying again, it only intermittently seems to recognise that I have an approval. ~ every second time it says I need an approval for the same contract.

woodenfurniture commented:
> On initial load, when the trade amount it 0, we don't get a gas estimate like we do with other swappers.
> 
> <img alt="Screenshot 2023-05-05 at 12 23 02 pm" width="473" src="https://user-images.githubusercontent.com/97164662/236365620-ad805c5b-7576-4819-a3f2-2befb421fa09.png">
> 
> When trading to or from ETH (which is not supported by 1inch), we show it as a possible swapper when the amount entered is 0. We should recognise that it's not a supported swapper at this point.
> 
> <img alt="Screenshot 2023-05-05 at 12 24 00 pm" width="472" src="https://user-images.githubusercontent.com/97164662/236365707-628515d1-ec3e-48c6-a42c-7dc384c81524.png">

ok for the sorting issue there were 2 bugs at play here:
1. the placeholder network fees were screwing with the swapper sorting. I've added a fix to handle unknown network fees
2. there was a race condition when we check the trading status of thorchain which was placing it at the end of the list in most cases (it was always slowest to resolve the promise). I've fixed to maintain the sorted order of swappers

![image](https://user-images.githubusercontent.com/125113430/236392828-de6b4f6e-5b46-4443-aed8-64852eff68c4.png)

0xApotheosis reviewed (commented):
Very tidy - swapper display and ordering issue is resolved.

Did you take a look at the approval issue?

woodenfurniture commented:
> Very tidy - swapper display and ordering issue is resolved.
> 
> Did you take a look at the approval issue?

I just successfully saw the approval work for 1inch but it took about 10 mins. I was recording my screen and stopped it after 5 mins but left the window open .. the trade confirmation appeared while i was looking in the console. I'll try get another one and keep the recording running

gomesalexandre reviewed (commented):
Tested locally:

- 1inch now shows unknown gas on default values for ETH pairs, which is better but still seems to be indicative that a swap would be possible to/from ETH? Obviously an edge case since I doubt many users would expand the swapper selection on 0-value, but definitely we should tackle from a heuristics point of view in the future, since it is *not* a supported swapper, it means the data is effectively wrong in this state

<img width="579" alt="image" src="https://user-images.githubusercontent.com/17035424/236432986-26b8436a-3f71-4aa7-9047-e6463b0ea83b.png">


- 1inch correctly shows unknown gas on a *valid* pair before entering an amount

<img width="576" alt="image" src="https://user-images.githubusercontent.com/17035424/236433148-4065c790-1588-4853-9768-dfa3d7fcab32.png">

- Expected amount is correct and matches the 1Inch quote

<img width="385" alt="image" src="https://user-images.githubusercontent.com/17035424/236433574-a9c84118-65b6-4d56-8cf3-25a5b08d899e.png">
<img width="497" alt="image" src="https://user-images.githubusercontent.com/17035424/236433658-2a935936-ca52-4b24-a36a-add3e75bdc33.png">

- As noted by @0xApotheosis , I also originally though that approval Txs are never picked, despite the polling working

<img width="622" alt="image" src="https://user-images.githubusercontent.com/17035424/236433983-32c84b35-5772-48ba-9a5c-9b30841166b2.png">

But it turns out they are, Oneinch API is just very slow to index these, which effectively looks like a bug since you'd need to wait at the very least a few min (possibly more, I had a debugger going through the polling, and started seeing happy responses after around ~5mn but I was going through the `approvalneeded` flow and inspecting it during that time so it could really be anything from a minute to many here)

<img width="612" alt="image" src="https://user-images.githubusercontent.com/17035424/236435086-9e41005d-6346-4394-991b-a322df5f4eb6.png">

Any reason why we don't use our regular approval needed logic i.e `getERC20Allowance`? This is what's used in all other swappers, and our node is going to be more reliable and faster to index approvals than an external API.

- "Miner Fees" aren't matching the actual gas cost in MM

<img width="384" alt="image" src="https://user-images.githubusercontent.com/17035424/236436235-8928d387-bad8-4b9d-8b64-b57785a4df9b.png">
<img width="545" alt="image" src="https://user-images.githubusercontent.com/17035424/236436266-141a9c4b-9242-42d2-9657-2a1cc11fd8d7.png">

I wasn't able to test the actual trade confirmation because of the gas cost discrepancy and not enough funds on test wallet, will get that monkey patched and tested

gomesalexandre commented on src/components/Trade/hooks/useAvailableSwappers.tsx:
```diff
@@ -158,19 +156,19 @@ export const useAvailableSwappers = () => {
         is actually supported by us, it's just currently halted.
        */
-      const swappersToDisplay = active.length > 0 ? active : halted
+      const active = swappersWithQuoteMetadataAndTradingStatus.filter(({ isActive }) => isActive)
+      const halted = swappersWithQuoteMetadataAndTradingStatus.filter(({ isActive }) => !isActive)
```
Can `lodash/partition` this if you feel like it, I got you @0xdef1cafe

gomesalexandre replied:
8cd5e3ab942e12bb4b6b24d653ddcc4fd4389178

gomesalexandre reviewed (commented):
Tested trade confirmation on my humble-sized test wallet, with the 1$ minimum (which is confirmed working) monkey patch:

<img width="620" alt="image" src="https://user-images.githubusercontent.com/17035424/236444396-9b1e779a-6e2b-47f2-821f-5942da57a16f.png">
<img width="623" alt="image" src="https://user-images.githubusercontent.com/17035424/236444405-cf760171-279e-4496-bd32-6c38b41df436.png">
<img width="432" alt="image" src="https://user-images.githubusercontent.com/17035424/236444420-c59761f0-d988-4e34-b5ad-16da8781786d.png">

- As noted previously, fees are currently over-estimated. Not a blocker for going live but definitely something to tackle as a direct follow-up after this

- Trade completion is happy and the amounts are matching the on-chain one

<img width="614" alt="Screenshot 2023-05-05 at 13 18 09" src="https://user-images.githubusercontent.com/17035424/236444516-3c6241ff-e25a-4a0e-a881-3aa23cbf8385.png">

<img width="538" alt="image" src="https://user-images.githubusercontent.com/17035424/236444587-9772dbc3-78c7-4fb7-a01b-1f2a0dc070b3.png">

- Not introduced by your PR, but the "ShapeShift Donation" becomes re-toggleable after Tx confirmation on-chain. Captured in https://github.com/shapeshift/web/issues/4438

- Txid is matching the on-chain one

<img width="1295" alt="image" src="https://user-images.githubusercontent.com/17035424/236445157-c1341b4b-b669-4259-874a-b095c3cbef85.png">

gomesalexandre commented on src/lib/swapper/swappers/OneInchSwapper/OneInchSwapper.ts:
```diff
@@ -71,6 +75,39 @@ export class OneInchSwapper implements Swapper<EvmChainId, true> {
+    if (sellAsset.chainId !== buyAsset.chainId) {
+      return Err(
+        makeSwapErrorRight({
+          message: '[getTradeQuote] cross chain swaps not supported',
```
noice

gomesalexandre commented on src/lib/swapper/swappers/OneInchSwapper/OneInchSwapper.ts:
```diff
@@ -71,6 +75,39 @@ export class OneInchSwapper implements Swapper<EvmChainId, true> {
+
+    if (
+      !isEvmChainId(chainId) ||
+      !isEvmChainId(sellAsset.chainId) ||
+      !isEvmChainId(buyAsset.chainId)
```
Given the previous check, we might not need this?

gomesalexandre replied:
@woodenfurniture FYI, not blocking but possibly superfluous

gomesalexandre reviewed (commented):
Retested after:

- https://github.com/shapeshift/web/pull/4433/commits/7abb363fbed2bb94facbc19e211b1c82647c22e1 confirmed using our node now properly detects approvals instantly
- https://github.com/shapeshift/web/pull/4433/commits/23610744a16c75316f5531b05e3f3c71b85ba758 confirmed this doesn't make 1Inch an available swapper anymore on ETH<->ERC-20 pairs

<img width="627" alt="image" src="https://user-images.githubusercontent.com/17035424/236448119-6c5380ee-89a1-4f10-8d17-e7ab37a1ab6c.png">
<img width="634" alt="image" src="https://user-images.githubusercontent.com/17035424/236448160-18757ef9-0726-4052-8758-880366216868.png">

gomesalexandre commented on src/components/Trade/Components/TradeQuotes/TradeQuote.tsx:
```diff
@@ -122,9 +122,9 @@ export const TradeQuoteLoaded: React.FC<TradeQuoteLoadedProps> = ({
-        .times(feeAssetFiatRate)
-        .toString()
+    ? bnOrZero(fromBaseUnit(quote.feeData.networkFeeCryptoBaseUnit, feeAsset.precision)).times(
+        feeAssetFiatRate,
```
heyyy ty

gomesalexandre commented on src/components/Trade/Components/TradeQuotes/TradeQuote.tsx:
```diff
@@ -218,7 +218,11 @@ export const TradeQuoteLoaded: React.FC<TradeQuoteLoadedProps> = ({
             <FaGasPump />
           </RawText>
-          <Amount.Fiat value={networkFeeFiat} />
+          {networkFeeFiat && !networkFeeFiat.isZero() ? (
+            <Amount.Fiat value={networkFeeFiat?.toString()} />
+          ) : (
+            translate('trade.unknownGas')
+          )}
```
This will break CoWSwap, where 0 is a valid miner fee (there are no ETH Tx fees for CoW, the fees are subtracted from the sell asset).

We might want to use different heuristics here, and check if the sell amount is currently zero in addition to this check.

Prod:

<img width="555" alt="image" src="https://user-images.githubusercontent.com/17035424/236450444-36b92b46-07db-445a-9a53-ef866cefa02b.png">

This diff:

<img width="562" alt="image" src="https://user-images.githubusercontent.com/17035424/236450556-8fbe3eff-b541-4c98-8dcb-d6689f8b99eb.png">
